File: very hard/Shadows of the Knight - Episode 2/Haskell/Shadows of the Knight - Episode 2.hs
import System.IO
import Control.Monad
import Data.List.Split (splitOn)
import Data.Maybe (fromMaybe)

-- Function narrowDimension:
--   Updates the search interval in one dimension
--   based on the feedback and the comparison between
--   the previous position (x0 or y0) and the current position (x or y).
--
-- For SAME, if (x0+x) (or y0+y) is even, the bomb is
-- exactly in the middle.
--
-- For WARMER/COLDER, we deduce which side of the interval to keep:
--   - If x > x0 and WARMER, the bomb is to the right of the middle,
--     so new_min = ((x0+x)/2) + 1.
--   - If x > x0 and COLDER, the bomb is to the left, so new_max = (x0+x)/2 - (if even) or = (x0+x)/2 (if odd).
--
-- The same reasoning applies for y.

narrowDimension :: Int -> Int -> Int -> Int -> String -> (Int, Int)
narrowDimension prev current minV maxV info
    | info == "SAME" =
        let sum' = prev + current
        in if even sum'
            then let middle = sum' `div` 2 in (middle, middle)
            else (minV, maxV)
    | info == "WARMER" =
        if current > prev
            then let sum' = prev + current
                     new_lower = sum' `div` 2 + 1
                 in (max minV new_lower, maxV)
            else if current < prev
                then let sum' = prev + current
                         new_upper = if even sum' then sum' `div` 2 - 1 else sum' `div` 2
                     in (minV, min maxV new_upper)
                else (minV, maxV)
    | info == "COLDER" =
        if current > prev
            then let sum' = prev + current
                     new_upper = if even sum' then sum' `div` 2 - 1 else sum' `div` 2
                 in (minV, min maxV new_upper)
            else if current < prev
                then let sum' = prev + current
                         new_lower = sum' `div` 2 + 1
                     in (max minV new_lower, maxV)
                else (minV, maxV)
    | otherwise = (minV, maxV)

-- Function narrow:
-- Applies narrowDimension on x as long as
-- the horizontal interval is not reduced to a single value,
-- then on y if x is already determined.

narrow :: Int -> Int -> Int -> Int -> Int -> Int -> Int -> Int -> String -> (Int, Int, Int, Int)
narrow x0 y0 x y x_min x_max y_min y_max info
    | x_min /= x_max =
        let (new_x_min, new_x_max) = narrowDimension x0 x x_min x_max info
        in (new_x_min, y_min, new_x_max, y_max)
    | y_min /= y_max =
        let (new_y_min, new_y_max) = narrowDimension y0 y y_min y_max info
        in (x_min, new_y_min, x_max, new_y_max)
    | otherwise = (x_min, y_min, x_max, y_max)

main :: IO ()
main = do
    hSetBuffering stdout NoBuffering -- DO NOT REMOVE
    [w, h] <- fmap (map read . words) getLine
    _ <- getLine -- max number of turns, unused
    [x0_init, y0_init] <- fmap (map read . words) getLine
    let x_min0 = 0
        x_max0 = w - 1
        y_min0 = 0
        y_max0 = h - 1
    gameLoop x0_init y0_init x0_init y0_init x_min0 x_max0 y_min0 y_max0 w h

gameLoop :: Int -> Int -> Int -> Int -> Int -> Int -> Int -> Int -> Int -> Int -> IO ()
gameLoop x0 y0 x y x_min x_max y_min y_max w h = do
    info <- getLine
    let (x_min', y_min', x_max', y_max') = narrow x0 y0 x y x_min x_max y_min y_max info

    let (x0', y0') = (x, y)
    (x', y') <- if x_min' /= x_max'
        then do
            let x1
                  | x0' == 0 && x_max' + x_min' + 1 /= w = (3 * x_min' + x_max') `div` 2 - x0'
                  | x0' == w - 1 && x_max' + x_min' + 1 /= w = (x_min' + 3 * x_max') `div` 2 - x0'
                  | otherwise = x_min' + x_max' - x0'
                x2 = if x1 == x0' then x0' + 1 else x1
                xFinal = max 0 (min (w - 1) x2)
            return (xFinal, y)
        else do
            let x1 = if x /= x_min' then x_min' else x
            yInput <- if x /= x_min'
                then do
                    putStrLn (show x1 ++ " " ++ show y)
                    getLine
                else return info
            let y0'' = y
            let y1 = if y_min' /= y_max'
                        then if y0'' == 0 && y_max' + y_min' + 1 /= h
                                then (3 * y_min' + y_max') `div` 2 - y0''
                             else if y0'' == h - 1 && y_max' + y_min' + 1 /= h
                                then (y_min' + 3 * y_max') `div` 2 - y0''
                             else y_max' + y_min' - y0''
                        else y_min'
                y2 = if y1 == y0'' then y0'' + 1 else y1
                yFinal = max 0 (min (h - 1) y2)
            return (x1, yFinal)

    putStrLn (show x' ++ " " ++ show y')
    gameLoop x0' y0' x' y' x_min' x_max' y_min' y_max' w h

File: very hard/Shadows of the Knight - Episode 2/Swift/Shadows of the Knight - Episode 2.swift
import Glibc
import Foundation

public struct StderrOutputStream: TextOutputStream {
    public mutating func write(_ string: String) {
        fputs(string, stderr)
    }
}
public var errStream = StderrOutputStream()

/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Function narrow_dimension:                                                                                      //
//   Updates the search interval in one dimension                                                                  //
//   based on the feedback and the comparison between                                                              //
//   the previous position (x0 or y0) and the current position (x or y).                                           //
//                                                                                                                 //
// For SAME, if (x0+x) (or y0+y) is even, the bomb is exactly in the middle.                                       //
// For WARMER/COLDER, we deduce which side of the interval to keep.                                                //
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
func narrowDimension(prev: Int, current: Int, min: Int, max: Int, info: String) -> (Int, Int) {
    let sum = prev + current
    var min = min
    var max = max

    switch info {
    case "SAME":
        if sum % 2 == 0 {
            let middle = sum / 2
            min = middle
            max = middle
        }
    case "WARMER":
        if current > prev {
            let newLower = (sum / 2) + 1
            if newLower > min {
                min = newLower
            }
        } else if current < prev {
            let newUpper = (sum % 2 == 0) ? (sum / 2 - 1) : (sum / 2)
            if newUpper < max {
                max = newUpper
            }
        }
    case "COLDER":
        if current > prev {
            let newUpper = (sum % 2 == 0) ? (sum / 2 - 1) : (sum / 2)
            if newUpper < max {
                max = newUpper
            }
        } else if current < prev {
            let newLower = (sum / 2) + 1
            if newLower > min {
                min = newLower
            }
        }
    default:
        break
    }

    return (min, max)
}

////////////////////////////////////////////////////////////////
// Function narrow:                                           //
// Applies narrow_dimension on x as long as                   //
// the horizontal interval is not reduced to a single value,  //
// then on y if x is already determined.                      //
////////////////////////////////////////////////////////////////
func narrow(x0: Int, y0: Int, x: Int, y: Int,
            xMin: Int, xMax: Int, yMin: Int, yMax: Int,
            info: String) -> (Int, Int, Int, Int) {
    print("narrow: x0: \(x0) y0: \(y0) x_min: \(xMin) x_max: \(xMax) y_min: \(yMin) y_max: \(yMax) info: \(info)", to: &errStream)

    var (xMin, xMax, yMin, yMax) = (xMin, xMax, yMin, yMax)

    if xMin != xMax {
        (xMin, xMax) = narrowDimension(prev: x0, current: x, min: xMin, max: xMax, info: info)
    } else if yMin != yMax {
        (yMin, yMax) = narrowDimension(prev: y0, current: y, min: yMin, max: yMax, info: info)
    }

    print("narrow : x_min: \(xMin) x_max: \(xMax) y_min: \(yMin) y_max: \(yMax)", to: &errStream)
    return (xMin, xMax, yMin, yMax)
}

let firstInput = readLine()!.split(separator: " ").map { Int($0)! }
let W = firstInput[0]
let H = firstInput[1]
let _ = Int(readLine()!)! // max turns, unused
let position = readLine()!.split(separator: " ").map { Int($0)! }
var x0 = position[0]
var y0 = position[1]
var x = x0
var y = y0

var xMin = 0, xMax = W - 1
var yMin = 0, yMax = H - 1

while true {
    var info = readLine()!

    (xMin, xMax, yMin, yMax) = narrow(x0: x0, y0: y0, x: x, y: y,
                                    xMin: xMin, xMax: xMax, yMin: yMin, yMax: yMax, info: info)

    print("narrow results: x_min: \(xMin) x_max: \(xMax) y_min: \(yMin) y_max: \(yMax)", to: &errStream)

    x0 = x
    y0 = y

    if xMin != xMax {
        if x0 == 0 && xMax + xMin + 1 != W {
            x = (3 * xMin + xMax) / 2 - x0
        } else if x0 == W - 1 && xMax + xMin + 1 != W {
            x = (xMin + 3 * xMax) / 2 - x0
        } else {
            x = xMax + xMin - x0
        }

        if x == x0 {
            x += 1
        }
        x = max(0, min(W - 1, x))
    } else {
        if x != xMin {
            x = xMin
            x0 = x
            print("\(x) \(y)")
            info = readLine()!
        }

        if yMin != yMax {
            if y0 == 0 && yMax + yMin + 1 != H {
                y = (3 * yMin + yMax) / 2 - y0
            } else if y0 == H - 1 && yMax + yMin + 1 != H {
                y = (yMin + 3 * yMax) / 2 - y0
            } else {
                y = yMax + yMin - y0
            }

            if y == y0 {
                y += 1
            }
            y = max(0, min(H - 1, y))
        } else {
            y = yMin
        }
    }

    print("\(x) \(y)")
}
